### PR TITLE
OLMIS-4383: Decreased number of db/external service calls

### PR DESCRIPTION
**Pull request by lukaslew**
This change is [Reviewable](https://reviewable.io/reviews/openlmis/openlmis-requisition/47)

### CODE REVIEW

File: src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java [hunk 1]
```diff
@@ -244,21 +248,122 @@ protected void checkIfPeriodIsValid(Requisition requisition) {
...
+
+    return profiler;
+  }
+
+  void stopProfiler(Profiler profiler, Object... exitArgs) {
+    profiler.stop().log();
+    extLogger.exit(exitArgs);
+  }
+
+  Requisition findRequisition(UUID requisitionId, Profiler profiler) {
```
Comment (palbecki) [inline src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java]: Can you remove code duplication from 3 methods below?

File: src/main/java/org/openlmis/requisition/web/RequisitionController.java [hunk 2]
```diff
@@ -222,47 +196,34 @@ public RequisitionDto initiate(@RequestParam(value = "program") UUID programId,
...
-          new Message(MessageKeys.ERROR_REQUISITION_NOT_FOUND, requisitionId));
-    }
-
-    profiler.start(VALIDATE_CAN_CHANGE_STATUS);
-    validateForStatusChange(requisition);
-
-    profiler.start(CHECK_IF_PERIOD_IS_VALID);
-    checkIfPeriodIsValid(requisition);
+    Requisition requisition = findRequisition(requisitionId, profiler);
+    checkPermission(profiler, () -> permissionService.canSubmitRequisition(requisitionId));
```
Comment (sebbrudzinski) [inline src/main/java/org/openlmis/requisition/web/RequisitionController.java]: All those canSubmit/canAuthorize etc. retrieve the requisition object from the database again. Do you think we can also improve this?

File: src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java (hunk 1 above)
Comment (lukaslew) [reply src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java]: I added one method that will handle this situation.

File: src/main/java/org/openlmis/requisition/web/RequisitionController.java (hunk 2 above)
Comment (lukaslew) [reply src/main/java/org/openlmis/requisition/web/RequisitionController.java]: I modified permission service and in most cases it does not retrieve requisition.

**Comment by chunky56** [comment]
Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java, line 289 at r3](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/47#-L8c6ONL1CVEouGz4cEO:-L8c6ONL1CVEouGz4cEP:b-2gyeaq) ([raw file](https://github.com/openlmis/openlmis-requisition/blob/046317db2c0c9bd79d388134c408673608ca96a8/src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java#L289)):*
> ```Java
>   }
> 
>   Requisition findRequisition(UUID requisitionId, Profiler profiler) {
> ```

I think we can improve performance here by calling exists to see if the entity exists, and throwing an error if not. If it does exist, then calling findOne and returning it. Same for facility and program below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/47)*

**Comment by chunky56** [comment]
Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java, line 289 at r3](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/47#-L8c6ONL1CVEouGz4cEO:-L8c8m7n8cVazTKZsrDP:b5go40l) ([raw file](https://github.com/openlmis/openlmis-requisition/blob/046317db2c0c9bd79d388134c408673608ca96a8/src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java#L289)):*
<details><summary><i>Previously, chunky56 (Chongsun Ahn) wrote…</i></summary><blockquote>

I think we can improve performance here by calling exists to see if the entity exists, and throwing an error if not. If it does exist, then calling findOne and returning it. Same for facility and program below.
</blockquote></details>

http://docs.openlmis.org/en/latest/conventions/performanceTips.html#avoid-loading-entities-unnecessarily

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/47)*

**Comment by chunky56** [comment]
Is there a reason why we are not doing this review on FishEye/Crucible?

**Comment by sebbrudzinski** [comment]
*[src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java, line 289 at r3](https://beta.reviewable.io/reviews/openlmis/openlmis-requisition/47#-L8c6ONL1CVEouGz4cEO:-L8cA9dtB7lNV8uNCByd:b-kyt5js) ([raw file](https://github.com/openlmis/openlmis-requisition/blob/046317db2c0c9bd79d388134c408673608ca96a8/src/main/java/org/openlmis/requisition/web/BaseRequisitionController.java#L289)):*
<details><summary><i>Previously, chunky56 (Chongsun Ahn) wrote…</i></summary><blockquote>

http://docs.openlmis.org/en/latest/conventions/performanceTips.html#avoid-loading-entities-unnecessarily
</blockquote></details>

I wonder if we would really be improving the performance doing that here. The most common case is that the requisition, facility and program exist and so you would be doing an additional call. We would only get an improvement for cases where requisition of the given ID doesn't exist (providing incorrect id on purpose) or if someone deletes program/facility that is used in the requisition. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/openlmis/openlmis-requisition/47)*

**Comment by lukaslew** [comment]
@chunky56 We do this by PR because of 3.3 release. From what I remember we decided to do this in this way.

**Pull request merged**